x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java

```diff
@@ -612,4 +630,216 @@ private static Expression propagate(Or or) {
         }
     }
+
+    /**
+     * Simplifies arithmetic expressions with BinaryComparisons and fixed point fields, such as: (int + 2) / 3 > 4 => int > 10
+     */
+    public static final class SimplifyComparisonsArithmetics extends
+        org.elasticsearch.xpack.ql.optimizer.OptimizerRules.OptimizerExpressionRule<EsqlBinaryComparison> {
+        BiFunction<DataType, DataType, Boolean> typesCompatible;
+
+        SimplifyComparisonsArithmetics(BiFunction<DataType, DataType, Boolean> typesCompatible) {
+            super(org.elasticsearch.xpack.ql.optimizer.OptimizerRules.TransformDirection.UP);
+            this.typesCompatible = typesCompatible;
+        }
+
+        @Override
+        protected Expression rule(EsqlBinaryComparison bc) {
+            // optimize only once the expression has a literal on the right side of the binary comparison
+            if (bc.right() instanceof Literal) {
+                if (bc.left() instanceof ArithmeticOperation) {
+                    return simplifyBinaryComparison(bc);
+                }
+                if (bc.left() instanceof Neg) {
+                    return foldNegation(bc);
+                }
+            }
+            return bc;
+        }
+
+        private Expression simplifyBinaryComparison(EsqlBinaryComparison comparison) {
+            EsqlArithmeticOperation operation = (EsqlArithmeticOperation) comparison.left();
+            BinaryArithmeticOperation function = operation.function();
+            // Modulo can't be simplified.
+            if (function.equals(MOD)) {
+                return comparison;
+            }
+            OperationSimplifier simplification = null;
+            if (isMulOrDiv(function)) {
+                simplification = new MulDivSimplifier(comparison);
+            } else if (function.equals(ADD) || function.equals(SUB)) {
+                simplification = new AddSubSimplifier(comparison);
+            }
+
+            return (simplification == null || simplification.isUnsafe(typesCompatible)) ? comparison : simplification.apply();
+        }
+
+        private static boolean isMulOrDiv(BinaryArithmeticOperation op) {
+            return op.equals(MUL) || op.equals(DIV);
+        }
+
+        private static Expression foldNegation(EsqlBinaryComparison bc) {
+            Literal bcLiteral = (Literal) bc.right();
+            Expression literalNeg = tryFolding(new Neg(bcLiteral.source(), bcLiteral));
+            return literalNeg == null ? bc : bc.reverse().replaceChildren(asList(((Neg) bc.left()).field(), literalNeg));
+        }
+
+        private static Expression tryFolding(Expression expression) {
+            if (expression.foldable()) {
+                try {
+                    expression = new Literal(expression.source(), expression.fold(), expression.dataType());
+                } catch (ArithmeticException | DateTimeException e) {
```

Comment:
I believe this is the source of the issue in https://github.com/elastic/elasticsearch/issues/108519, as ES|QL will never throw from folding.  It's not as simple as just checking if we got a non-null value in the literal here though, as generating the null literal will have put the warning on the stack already.